src/runtime/virtcontainers/stratovirt.go

```diff
@@ -0,0 +1,1306 @@
+	params = append(params, "-device", fmt.Sprintf("%s,%s", driver, strings.Join(SerializeParams(devParams, "="), ",")))
+	return params
+}
+
+type consoleDevice struct {
+	driver   VirtioDriver
+	id       string
+	backend  string
+	charID   string
+	devType  string
+	charDev  string
+	deviceID string
+}
+
+func (c consoleDevice) getParams(config *StratovirtConfig) []string {
+	var params []string
+	var devParams []Param
+	var charParams []Param
+	var conParams []Param
+
+	driver := consoleDriver[c.driver]
+	if c.id != "" {
+		devParams = append(devParams, Param{"id", c.id})
+	}
+
+	conParams = append(conParams, Param{"chardev", c.charDev})
+	conParams = append(conParams, Param{"id", c.deviceID})
+	params = append(params, "-device", fmt.Sprintf("%s,%s", driver, strings.Join(SerializeParams(devParams, "="), ",")))
+
+	charParams = append(charParams, Param{"id", c.charID})
+	charParams = append(charParams, Param{"path", config.consolePath})
+	params = append(params, "-chardev", fmt.Sprintf("%s,%s,server,nowait", c.backend, strings.Join(SerializeParams(charParams, "="), ",")))
+	params = append(params, "-device", fmt.Sprintf("%s,%s,nr=0", c.devType, strings.Join(SerializeParams(conParams, "="), ",")))
+	return params
+}
+
+// StratovirtConfig keeps the custom settings and parameters to start virtual machine.
+type StratovirtConfig struct {
+	name          string
+	uuid          string
+	machineType   string
+	vmPath        string
+	smp           uint32
+	memory        uint64
+	kernelPath    string
+	params        string
+	rootfsPath    string
+	initrdPath    string
+	devices       []VirtioDev
+	qmpSocketPath govmmQemu.QMPSocket
+	consolePath   string
+	fsSockPath    string
+	fds           []*os.File
+}
+
+func (config *StratovirtConfig) appendFDs(fds []*os.File) []int {
+	var fdInts []int
+
+	oldLen := len(config.fds)
+
+	config.fds = append(config.fds, fds...)
+
+	// The magic 3 offset comes from https://golang.org/src/os/exec/exec.go:
+	//     ExtraFiles specifies additional open files to be inherited by the
+	//     new process. It does not include standard input, standard output, or
+	//     standard error. If non-nil, entry i becomes file descriptor 3+i.
+	// This means that arbitrary file descriptors fd0, fd1... fdN passed in
+	// the array will be presented to the guest as consecutive descriptors
+	// 3, 4... N+3. The golang library internally relies on dup2() to do
+	// the renumbering.
+	for i := range fds {
+		fdInts = append(fdInts, oldLen+3+i)
+	}
+
+	return fdInts
+}
+
+// State keeps StratoVirt device and pids state.
+type State struct {
+	mmioBlkSlots [maxMmioBlkCount]bool
+	pid          int
+	virtiofsPid  int
+}
+
+type stratovirt struct {
+	id             string
+	path           string
+	ctx            context.Context
+	fds            []*os.File
+	config         HypervisorConfig
+	qmpMonitorCh   qmpChannel
+	svConfig       StratovirtConfig
+	state          State
+	stopped        int32
```

Comment:
Is there any reason not to use the `atomic.Bool` type here? I can only see `stopped` set to 1 and compared to 0 so it looks like it's used as a boolean anyway.  If that's the case the bool type would communicate the field's semantics better.  Seeing it as an integer makes one wonder if there are more than two states of being stopped.

Reply:
I think this is a good suggestion! The semantics of this variable is just a Boo value, such modification can reduce the ambiguity.

I'll take note of it and make batch changes with other further suggestions. :）